src/translate.py

```diff
@@ -3832,36 +3871,80 @@ def global_decls(self, fmt: Formatter) -> str:
                 else:
                     qualifier = "extern"
 
-                if type.is_function():
+                if sym.type.is_function():
                     comments.append(qualifier)
                     qualifier = ""
 
-                # Try to convert the data from .data/.rodata into an initializer
-                entry = self.asm_data.values.get(name)
-                if entry is not None:
-                    if entry.is_readonly:
-                        comments.append("const")
-                    data_size = entry.size_bytes()
-                    type_size_bits = type.get_size_bits()
-                    if type_size_bits and data_size != type_size_bits // 8:
-                        # TODO: Use heuristic to detect padding
-                        type_size = type_size_bits // 8
-                        if data_size // type_size > 1:
-                            comments.append(f"array: [{data_size // type_size}]")
+                # Try to guess the symbol's `array_dim` if we have a data entry for it,
+                # and it does not exist in the typemap (`not is_global`) or dim is unknown
+                # (Otherwise, if the dim is provided by the typemap, we trust it.)
+                if sym.asm_data_entry and (not is_global or sym.array_dim == 0):
+                    assert sym.array_dim is None or sym.array_dim == 0
+                    # The size of the data entry is uncertain, because of padding
+                    # between sections. Generally `(max_data_size - data_size) < 16`.
+                    data_size, max_data_size = sym.asm_data_entry.size_range_bytes()
```

Comment:
rename data_size -> min_data_size, and create a data_size member further down just above where it's used